packages/flutter/lib/src/rendering/sliver_fill.dart

```diff
@@ -114,34 +115,80 @@ class RenderSliverFillRemaining extends RenderSliverSingleBoxAdapter {
   }) : assert(hasScrollBody != null),
        super(child: child);
 
-  /// Whether the child has a scrollable body, this value cannot be null.
+  /// Indicates whether the child has a scrollable body, this value cannot be
+  /// null.
   ///
   /// Defaults to true such that the child will extend beyond the viewport and
   /// scroll, as seen in [NestedScrollView].
   ///
   /// Setting this value to false will allow the child to fill the remainder of
-  /// the viewport and not extend further.
+  /// the viewport and not extend further. However, if the
+  /// [precedingScrollExtent] exceeds the size of the viewport, the sliver will
+  /// defer to the child's size rather than overriding it.
   bool hasScrollBody;
 
+  /// Indicates whether the child should stretch to fill the overscroll area
+  /// created by certain scroll physics, such as iOS' default scroll physics.
+  /// This value cannot be null. This flag is only relevant when the
+  /// [hasScrollBody] value is false.
+  ///
+  /// Defaults to false, meaning the default behavior is for the child to
+  /// maintain its size and not extend into the overscroll area.
+  bool fillOverscroll;
+
   @override
   void performLayout() {
-    final double extent = constraints.remainingPaintExtent
-      - math.min(constraints.overlap, 0.0)
-      // Adding the offset for when this SliverFillRemaining is not scrollable,
-      // so it will stretch to fill on overscroll.
-      + (hasScrollBody ? 0.0 : constraints.scrollOffset);
-    if (child != null)
-      child.layout(constraints.asBoxConstraints(minExtent: extent, maxExtent: extent), parentUsesSize: true);
+    double childExtent;
+    double extent = constraints.viewportMainAxisExtent - constraints.precedingScrollExtent;
+    double maxExtent = constraints.remainingPaintExtent - math.min(constraints.overlap, 0.0);
+
+    if (hasScrollBody) {
+      extent = maxExtent;
+      if (child != null)
+        child.layout(
+          constraints.asBoxConstraints(
+            minExtent: extent,
+            maxExtent: extent,
+          ),
+          parentUsesSize: true,
+        );
+    } else if (child != null) {
+      child.layout(constraints.asBoxConstraints(), parentUsesSize: true);
+
+      switch (constraints.axis) {
+        case Axis.horizontal:
+          childExtent = child.size.width;
+          break;
+        case Axis.vertical:
+          childExtent = child.size.height;
+          break;
+      }
+
+      if (constraints.precedingScrollExtent > constraints.viewportMainAxisExtent || childExtent > extent)
+        extent = childExtent;
+      if (maxExtent < extent)
+        maxExtent = extent;
+      if ((fillOverscroll ? maxExtent : extent) > childExtent) {
+        child.layout(
+          constraints.asBoxConstraints(
+            minExtent: extent,
+            maxExtent: fillOverscroll ? maxExtent : extent,
+          ),
+          parentUsesSize: true,
+        );
+      }
+    }
+
+    assert(extent.isFinite);
```

Comment:
Let's add a test covering this, and ideally have a message here that explains what can be done if anything (e.g. wrapping the child in something else or changing the child to some other type?).

Reply:
(post-offline chat) I haven't been able to trigger this assert in a test, as it is currently covered by the child's layout method. I will leave this assertion in place thought o guard against future changes that may result in triggering this assert.
I have added a message to explain why this assertion would fail.